cli/core/test/e2e/framework/plugin_operations.go

```diff
@@ -0,0 +1,299 @@
+	if err != nil {
+		return pluginsMeta.pluginLocalPath, err
+	}
+	defer of.Close()
+	yamlData, err := yaml.Marshal(&pluginOverlayObj)
+	if err != nil {
+		return pluginsMeta.pluginLocalPath, err
+	}
+	err = os.WriteFile(overlayFile, yamlData, 0644)
+	if err != nil {
+		return pluginsMeta.pluginLocalPath, err
+	}
+	return pluginsMeta.pluginLocalPath, nil
+}
+
+// generatePluginBinary creates the script based plugin binary file for given plugin metadata, saves in local file system
+func (po *localOCIPluginOps) generatePluginBinary(pm *PluginMeta) (string, error) {
+	nameWithTarget := pm.target + "_" + pm.name
+	pm.pluginBinaryFileName = nameWithTarget + "-" + pm.os + "-" + pm.version
+	if err := CreateDir(pm.pluginLocalPath); err != nil {
+		return "", err
+	}
+	filePath := filepath.Join(pm.pluginLocalPath, (pm.pluginBinaryFileName))
+	f, err := os.OpenFile(filePath, os.O_CREATE|os.O_WRONLY, 0755)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	fmt.Fprintf(f, ScriptBasedPluginTemplate, pm.name, pm.target, pm.description, pm.version,
+		pm.sha, pm.group, strconv.FormatBool(pm.hidden), pm.aliases, pm.version, pm.name, pm.name)
+	return filePath, nil
+}
+
+// CreateDir creates directory if not exists
+func CreateDir(dir string) error {
+	err := os.MkdirAll(dir, 0750)
+	if err != nil && !os.IsExist(err) {
+		return err
+	}
+	return nil
+}
+
+// generatePluginDiscoveryOverlay creates plugin overly object for given plugin metadata
+func (po *localOCIPluginOps) generatePluginDiscoveryOverlay(pm *PluginMeta) (plugin *CLIPlugin, err error) {
+	plugin = &CLIPlugin{}
+	plugin.TypeMeta.Kind = "CLIPlugin"
+	plugin.TypeMeta.APIVersion = "cli.tanzu.vmware.com/v1alpha1"
+	plugin.Metadata.Name = pm.name
+
+	var artifactsMap = make(map[string]plugintypes.ArtifactList)
+	artifacts := make([]plugintypes.Artifact, 1)
+	artifacts[0].OS = pm.os
+	artifacts[0].Image = pm.binaryDistributionURL
+	artifacts[0].Arch = pm.arch
+	artifacts[0].Type = pm.discoveryType
+
+	artifactsMap[pm.version] = artifacts
+	plugin.Spec.Artifacts = artifactsMap
+	plugin.Spec.Description = pm.description
+	plugin.Spec.Optional = pm.optional
+	plugin.Spec.RecommendedVersion = pm.version
+
+	return plugin, nil
+}
+
+func (po *localOCIPluginOps) AddPluginDiscoverySource(discoveryOpts *DiscoveryOptions) error {
+	addCmd := fmt.Sprintf(AddPluginSource, discoveryOpts.Name, discoveryOpts.SourceType, discoveryOpts.URI)
+	_, _, err := po.cmdExe.Exec(addCmd)
+	return err
+}
+
+func (po *localOCIPluginOps) ListPlugins() ([]PluginListInfo, error) {
+	out, _, err := po.cmdExe.Exec(ListPlugins)
+	if err != nil {
+		return nil, err
+	}
+	var list []PluginListInfo
+	outStr := out.String()
+	start := strings.LastIndex(outStr, "[\n")
+	end := strings.LastIndex(outStr, "]")
+	jsonData := outStr[start : end+1]
+	err = yaml.Unmarshal([]byte(jsonData), &list)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to construct yaml node from config get output")
+	}
```

Comment:
considering the output is JSON, we should be using JSON unmarshal instead of YAML unmarshal. That will eliminate the string processing-related code.
Please refer to this example: https://go.dev/play/p/EApTwngSCKx